15/umbraco-cms/tutorials/inspiration/migrating-macros/migrating-macros.md

```diff
@@ -0,0 +1,889 @@
+    {
+        if (TargetBlockContentTypeKey == Guid.Empty)
+        {
+            // can't process as the doctype is not present
+            return originalValue;
+        }
+        
+        // this migrator assumes that the conversion from old RTE values to new (markup being wrapped in json) has already been completed.
+        RichTextPropertyEditorHelper.TryParseRichTextEditorValue(originalValue, _jsonSerializer, _logger,
+            out var richTextEditorValue);
+
+        if (richTextEditorValue == null)
+        {
+            return originalValue;
+        }
+        
+        // collect the values by using a group regex, read around possible downsides at the regex definition
+        var macroMatches = MacroTagRegex.Matches(richTextEditorValue.Markup);
+        foreach (Match? macroMatch in macroMatches)
+        {
+            if (macroMatch == null)
+            {
+                continue;
+            }
+         
+            // every macro needs its values assigned to a comparable blockValue
+            var blockKey = Guid.NewGuid();
+            richTextEditorValue.Blocks ??= new RichTextBlockValue();
+            // add the block to the inline contentData
+            richTextEditorValue.Blocks.ContentData.Add(new BlockItemData
+            {
+                Key = blockKey,
+                ContentTypeKey = TargetBlockContentTypeKey,
+                ContentTypeAlias = BlockContentTypeAlias,
+                Values = new List<BlockPropertyValue>
+                {
+                    CreateInvariantTextboxBlockPropertyValue("title", macroMatch.Groups["title"].Value),
+                    CreateInvariantTextboxBlockPropertyValue("youtubeVideoId", macroMatch.Groups["youtubeVideoId"].Value)
+                }
+            });
+            // expose the contentblock as an invariant
+            richTextEditorValue.Blocks.Expose.Add(new BlockItemVariation
+            {
+                ContentKey = blockKey,
+                Culture = null,
+                Segment = null,
+            });
+            // setup the layout
+            var layoutItem = new RichTextBlockLayoutItem(blockKey);
+            var layoutList = richTextEditorValue.Blocks.Layout.ContainsKey(Constants.PropertyEditors.Aliases.RichText)
+                ? richTextEditorValue.Blocks.Layout[Constants.PropertyEditors.Aliases.RichText].ToList()
+                : new List<IBlockLayoutItem>();
+            layoutList.Add(layoutItem);
+            richTextEditorValue.Blocks.Layout[Constants.PropertyEditors.Aliases.RichText] = layoutList;
+            
+            // now that we have converted the data into a block, replace the macro tag by a block tag
+            richTextEditorValue.Markup = richTextEditorValue.Markup.ReplaceFirst(macroMatch.Value, $"<umb-rte-block-inline data-content-key=\"{blockKey}\"></umb-rte-block-inline>");
+        }
+        
+        return _jsonSerializer.Serialize(richTextEditorValue);
```

Comment:
Use `RichTextPropertyEditorHelper.SerializeRichTextEditorValue()` to serialize the RTE content, so we're doing exactly the same as the CMS (and Deploy for that matter).